pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go

```diff
@@ -0,0 +1,334 @@
+	ErrorInjectionCh chan error
+}
+
+// ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.
+func (k TestingKnobs) ModuleTestingKnobs() {}
+
+var _ base.ModuleTestingKnobs = (*TestingKnobs)(nil)
+
+// NewWatcher instantiates a Watcher. The two key drivers of the watcher are
+// translateEvent and onUpdate.
+//
+// The translateEvent function is called on each RangeFeedValue event as they
+// arrive (including the events synthesized from the initial scan). Callers may
+// utilize translateEvent to trigger actions without waiting for the delay due
+// to resolving the complete set of spans. If nil is returned, no event will
+// be buffered.
+//
+// Once a resolved timestamp for the complete set of spans has been processed,
+// an Update is passed to onUpdate. Note that the update may contain no
+// buffered events.
+//
+// Callers can control whether the values passed to translateEvent carry a
+// populated PrevValue using the withPrevValue parameter. See
+// rangefeed.WithDiff for more details.
+func NewWatcher(
+	name redact.SafeString,
+	clock *hlc.Clock,
+	rangeFeedFactory *rangefeed.Factory,
+	bufferSize int,
+	spans []roachpb.Span,
+	withPrevValue bool,
+	translateEvent TranslateEventFunc,
+	onUpdate OnUpdateFunc,
+	knobs *TestingKnobs,
+) *Watcher {
+	w := &Watcher{
+		name:             name,
+		clock:            clock,
+		rangefeedFactory: rangeFeedFactory,
+		spans:            spans,
+		bufferSize:       bufferSize,
+		withPrevValue:    withPrevValue,
+		translateEvent:   translateEvent,
+		onUpdate:         onUpdate,
+	}
+	if knobs != nil {
+		w.knobs = *knobs
+	}
+	return w
+}
+
+// Start calls Run on the Watcher as an async task with exponential backoff.
+// If the Watcher ran for what is deemed long enough, the backoff is reset.
+func Start(ctx context.Context, stopper *stop.Stopper, c *Watcher, onError func(error)) error {
+	return stopper.RunAsyncTask(ctx, string(c.name), func(ctx context.Context) {
+		ctx, cancel := stopper.WithCancelOnQuiesce(ctx)
+		defer cancel()
+
+		const aWhile = 5 * time.Minute // arbitrary but much longer than a retry
+		for r := retry.StartWithCtx(ctx, base.DefaultRetryOptions()); r.Next(); {
+			started := timeutil.Now()
+			if err := c.Run(ctx); err != nil {
+				if errors.Is(err, context.Canceled) {
+					return // we're done here
+				}
+				if onError != nil {
+					onError(err)
+				}
+
+				if timeutil.Since(started) > aWhile {
+					r.Reset()
+				}
+
+				log.Warningf(ctx, "%s: failed with %v, retrying...", c.name, err)
+				continue
+			}
+
+			return // we're done here (the stopper was stopped, Run exited cleanly)
+		}
+	})
+}
+
+// Run establishes a rangefeed over the global store of span configs.
+// This is a blocking operation, returning only when the context is canceled,
+// or when a retryable error occurs. For the latter, it's expected that callers
+// will re-run the watcher.
+func (s *Watcher) Run(ctx context.Context) error {
```

Comment:
Does this need to be exported?

Reply:
Also, refers to "global store of span configs".

Reply:
I sort of like that it's exported, it makes testing other things useful, and it had clear semantics: run synchronously until there's a permanent error.